File: 202410/438.Find All Anagrams in a String.cpp
/*
 * @lc app=leetcode id=438 lang=cpp
 * @lcpr version=20001
 *
 * [438] Find All Anagrams in a String
 */

// @lcpr-template-start
using namespace std;
#include <algorithm>
#include <array>
#include <bitset>
#include <climits>
#include <deque>
#include <functional>
#include <iostream>
#include <list>
#include <queue>
#include <stack>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
// @lcpr-template-end
// @lc code=start
class Solution
{
public:
    vector<int> findAnagrams(string s, string p)
    {
        unordered_map<char, int> window;
        unordered_map<char, int> need;
        vector<int> res;

        // init need
        for (auto &&c : p)
        {
            need[c]++;
        }

        // substring
        int left = 0, right = 0;
        int valid = 0;

        while (right < s.size())
        {
            auto c = s[right];
            right++;

            if(need.count(c)){
                window[c]++;
                if (window[c] == need[c])
                {
                    valid++;
                }
            }
            else{
                window.clear();
                valid = 0;
                left = right;
            }

            while (right - left >= p.size())
            {
                if (valid == need.size())
                {
                    res.push_back(left);
                }

                auto d = s[left];
                left++;
                if(need.count(d)){
                    if (window[d] == need[d])
                    {
                        valid--;
                    }
                    window[d]--;
                }
            }
        }
        return res;
    }
};
// @lc code=end

/*
// @lcpr case=start
// "cbaebabacd"\n"abc"\n
// @lcpr case=end

// @lcpr case=start
// "abab"\n"ab"\n
// @lcpr case=end

 */

File: 202503/438.Find All Anagrams in a String.java
/*
 * @lc app=leetcode id=438 lang=java
 * @lcpr version=30103
 *
 * [438] Find All Anagrams in a String
 */

import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;

// @lc code=start
class Solution {

    public List<Integer> findAnagrams(String s, String p) {
        HashMap<Character, Integer> targetMap = new HashMap<>();
        List<Integer> res = new ArrayList<>();

        for (int i = 0; i < p.length(); i++) {
            char c = p.charAt(i);
            targetMap.put(c, targetMap.getOrDefault(c, 0) + 1);
        }

        HashMap<Character, Integer> windowMap = new HashMap<>();
        int need = 0;

        // [lo, hi]
        int lo = 0, hi = 0;
        while (hi < s.length()) {
            char c = s.charAt(hi);
            hi++;
            if (targetMap.containsKey(c)) {
                int count = windowMap.getOrDefault(c, 0) + 1;
                windowMap.put(c, count);
                if (count == targetMap.get(c)) {
                    need++;
                }
            }

            while (hi - lo >= p.length()) {
                if (need == targetMap.size()) {
                    res.add(lo);
                }

                char d = s.charAt(lo);
                lo++;
                if (targetMap.containsKey(d)) {
                    int count = windowMap.get(d);
                    if (count == targetMap.get(d)) {
                        need--;
                    }
                    windowMap.put(d, count - 1);
                }

            }
        }

        return res;
    }
}
// @lc code=end

/*
// @lcpr case=start
// "cbaebabacd"\n"abc"\n
// @lcpr case=end

// @lcpr case=start
// "abab"\n"ab"\n
// @lcpr case=end

 */
